algorithms/pagerank.ddR/R/dpagerank.R

```diff
@@ -0,0 +1,280 @@
+
+
+  niter <- as.numeric(niter)
+  if (! niter > 0 )
+    stop("Invalid iteration count, it should be a positive number")
+  eps <- as.numeric(eps)
+  if (! eps > 0 )
+    stop("Invalid value for 'eps', it should be a positive number")
+  damping <- as.numeric(damping)
+  if (damping <= 0 || damping >= 1)
+    stop("Invalid damping factor, it should be between 0 and 1")
+
+  if (trace) {    # end of timing step
+    endtime <- proc.time()
+    spentTime <- endtime-starttime
+    cat("Spent time:",(spentTime)[3],"sec\n")
+  }
+
+  ### Initialization
+  if (trace) {
+    cat("Initialization PR and OutDegree\n")
+    starttime<-proc.time()
+  }
+  maxdiff <- eps
+  # initial values for PR is 1/nVertices like most literature (it is 1-damping in igraph package)
+  PR <- dmapply(function(data, nrow, ncol) matrix(data=data, nrow=nrow, ncol=ncol), 
+                ncol=psize(dgraph)[,2], MoreArgs=list(nrow=1, data=1-damping), output.type="darray", combine="cbind")
+
+  if (trace) {    # end of timing step
+    endtime <- proc.time()
+    spentTime <- endtime-starttime
+    cat("Spent time:",(spentTime)[3],"sec\n")
+  }
+  
+
+  #Let's get the number of outgoing edges in each partition
+  if (trace) {
+    cat("Calculating the number of outgoing edges\n")
+    starttime<-proc.time()
+  }
+
+  if (is.null(weights)) { # when there is no weight on the edges
+    OutDegree <- dmapply(function(X) X, rowSums(dgraph), output.type="darray", nparts=c(1, 1), combine='rbind') 
+  } else {  # when there are weights on the edges
+    weightedgraph <- dmapply(function(DGi, Wi) DGi * Wi, parts(dgraph), parts(weights), output.type=dgraph@type, nparts=nparts(dgraph), combine='rbind')
+    OutDegree <- dmapply(function(X) X, rowSums(weightedgraph), output.type="darray", nparts=c(1, 1), combine='rbind')
+  }
+
+  if (trace) {    # end of timing step
+    endtime <- proc.time()
+    spentTime <- endtime-starttime
+    cat("Spent time:",(spentTime)[3],"sec\n")
+  }
+
+
+  ### iterations
+  iteration_counter <- 1
+  if (trace)  iterations_start <- proc.time()
+  while (niter >= iteration_counter && maxdiff >= eps) {
+    if (trace) {
+      cat("Iteration: ",iteration_counter,"\n")
+      starttime <- proc.time()
+    }
+    iteration_counter <- iteration_counter + 1
+
+    # the new PR (PageRank vector) is calculated in each iteration
+    if(is.null(weights) && is.null(personalized)) {
+        PR_new <- dmapply(function(dg, prOld, damping, TPs, peri, wi) {
+                        if(class(dg) == "matrix") {
+                            .Call("pagerank_vm", prOld, dg, TPs, damping, NULL, NULL, PACKAGE="pagerank.ddR")
+                        } else {
+                            .Call("pagerank_spvm", prOld, dg, TPs, damping, NULL, NULL, PACKAGE="pagerank.ddR")
+                        }                        
+                    }, dg=parts(dgraph), 
+                    MoreArgs=list(prOld=PR, damping=damping, TPs=OutDegree), output.type="darray", combine="rbind")
```

Comment:
Well, it was working as expected although not very intuitive. PR_new must have a single row, but the C function used to return a vector. Then, the vectors of different partitions used to be converted to a matrix with a single column. Therefore, as a trick I used combine="rbind" to make them single row instead.
I agree that it makes the code unreadable; hence, I modified the C function to return a single row matrix, and made it combine="cbind".